br/pkg/lightning/lightning.go

```diff
@@ -331,7 +337,8 @@ func (l *Lightning) run(taskCtx context.Context, taskCfg *config.Config, o *opti
 
 	utils.LogEnvVariables()
 
-	ctx, cancel := context.WithCancel(taskCtx)
+	ctx := metric.NewContext(taskCtx, l.metrics)
```

Comment:
what's the specific problem that we don't choose to add a WithMetricFactory to RunOnceWithOptions?

Reply:
When Lightning runs in server mode, multiple tasks need to share the same metrics. So metrics must be created before `run` and stored in the  Lightning struct. I hope server mode and lightning being used as library could share the same logic, so this PR didn't add new options to `RunOnceWithOptions`.

Reply:
maybe if there is not Option for metrics, we use metric.NewMetrics(NewDefaultFactory) and register it, otherwise we use ones from Options. Not a strong request.

Reply:
There is another custom registry `PromRegistry`. In this way, it's hard to determine when to register metrics.

Reply:
In lightning's server mode, as we only allow one task to run at one time, I think it's ok to bind metrics to task. Metrics across multi task is meaningless and the statistics based on this(such as global progress log) are wrong.

Reply:
It sounds like a bug that was not noticed before. I will move these logic to `RunOnceWithOptions`.

Reply:
Fixed in https://github.com/pingcap/tidb/pull/35127/commits/402c686639c1bff2b538d9a0cfb5aee1e6a9de43.

Reply:
@glorv @lance6716 PTAL again, thanks.

Reply:
lgtm